### PR TITLE
fix: helps incremental builds of large datasets by paginating

### LINKED ISSUE #149: Build delta change computation does not scale to large datasets
The [current delta logic](https://github.com/sanity-io/gatsby-source-sanity/blob/main/src/gatsby-node.ts#L255-L257) relies on getting all of the IDs from a dataset first. However, with large datasets, this can take a very long time, and might even time out, as the query in question loads all the IDs in one go:

```groq
*[!(_type match "system.**")]._id
```

This will effectively cause all IDs to be loaded and rendered into JSON before it is returned to the client.

A more incremental approach is needed to avoid running into this issue, where the IDs are paginated:

```groq
*[!(_type match "system.**") && id > $prevId][0..1000]._id
```

For each page, `prevId` can be filled in with the highest ID encountered. This is much faster, and avoids running into timeouts.

An even better approach would be to only fetch anything with `_updatedAt > $lastBuildTime`, though this does not solve how to sync deleted documents. But it's possible that the history API could be used for this.

There are other aspects of this logic that could be optimized for large datasets. For example, `documentIds.includes()` is called for every single document, which is a linear search that's going to be very slow if the document ID list contains thousands of documents. Instead, the IDs should be indexed into an efficient set first for fast lookup.

### PR DESCRIPTION
**Pull request by atombender**
Helps incremental builds of large datasets by paginating fetching of document IDs, and optimizing their lookups using a set. Fixes #149.

### CODE REVIEW

File: src/gatsby-node.ts [hunk 1]
```diff
@@ -190,6 +190,32 @@ export const createSchemaCustomization: GatsbyNode['createSchemaCustomization']
...
+    const batch = await client.fetch<string[]>(
+      prevId !== undefined
+        ? `*[!(_type match "system.**") && _id > $prevId][0...$batchSize]._id`
+        : `*[!(_type match "system.**")][0...$batchSize]._id`,
+      {
+        prevId: prevId || null,
+        batchSize,
+      },
+    )
+    if (batch.length === 0) {
```
Comment (runeb) [inline src/gatsby-node.ts:209]: Could you save one query by changing this to
```
if (batch.length < batchSize) {
```
?

Comment (atombender) [reply src/gatsby-node.ts]: I guess. I don't like to make assumptions about when the end is reached, but I suppose in this case there'll never be any variation in the slicing. 👍

**Comment by atombender** [comment]
Thoughts on merging and releasing this, @runeb? Who else might want to take a look?

**Comment by AbdallahAbis** [comment]
Any updates on this?

**Comment by atombender** [comment]
@AbdallahAbis Waiting for ecosystem team to sign off on this.

**Comment by AbdallahAbis** [comment]
Any updates!

**Comment by atombender** [comment]
No, see above; I need sign-off from the team that owns this plugin. If this is urgent, perhaps you could publish the PR as a temporary NPM package under an org that you control?

**Comment by AbdallahAbis** [comment]
Maybe no team owns this plugin...

**Review by stipsan — APPROVED** [review]
Looking good 👍

**Comment by atombender** [comment]
@AbdallahAbis Released as 7.4.2.

**Comment by AbdallahAbis** [comment]
@atombender Thanks!

**Pull request merged**